File: 2021/09/27/ShortestCompletingWord.java
/**
 * https://leetcode.com/problems/shortest-completing-word/
 * @author Andrew Jarombek
 * @since 9/27/2021
 */

import java.util.HashMap;
import java.util.Map;

public class ShortestCompletingWord {
    public String shortestCompletingWord(String licensePlate, String[] words) {
        String lp = licensePlate.toLowerCase();
        Map<Character, Integer> lpMap = new HashMap<>();

        for (char c : lp.toCharArray()) {
            lpMap.put(c, lpMap.containsKey(c) ? lpMap.get(c) + 1 : 1);
        }

        int mostMatches = -1;
        String result = null;

        for (String word : words) {
            int matches = 0;
            Map<Character, Integer> wordMap = new HashMap<>();

            for (char c : word.toCharArray()) {
                Integer lpCount = lpMap.get(c);
                Integer wordCount = wordMap.get(c);
                wordCount = wordCount == null ? 0 : wordCount;

                if (lpCount != null && lpCount > wordCount) {
                    matches++;
                    wordMap.put(c, wordCount + 1);
                }
            }

            if (matches > mostMatches || (matches == mostMatches && word.length() < result.length())) {
                mostMatches = matches;
                result = word;
            }
        }

        return result;
    }
}

File: 2022/04/19/shortestCompletingWord.ts
/**
 * https://leetcode.com/problems/shortest-completing-word/
 * @author Andrew Jarombek
 * @since 4/19/2022
 */

function shortestCompletingWord(licensePlate: string, words: string[]): string {
    let lpMap = new Map<string, number>();

    for (let c of licensePlate) {
        c = c.toLowerCase();

        if (c >= 'a' && c <= 'z')
            lpMap.set(c, (lpMap.get(c) ?? 0) + 1);
    }

    let result = "";
    for (let word of words) {
        let charMap = new Map<string, number>();

        for (let c of word) {
            charMap.set(c, (charMap.get(c) ?? 0) + 1);
        }

        let valid = true;
        for (let [key, value] of lpMap.entries()) {
            let count = charMap.get(key) ?? 0;
            if (count == 0 || value > count) {
                valid = false;
                break;
            }
        }

        if (valid && (!result || result.length > word.length)) {
            result = word;
        }
    }

    return result;
}